docs/enhancements/config-image-versions.md

```diff
@@ -0,0 +1,143 @@
+---
+title: config-image-versions
+authors:
+  - "@lranjbar"
+creation-date: 2021-10-27
+last-updated: 2021-10-29
+---
+
+# Configurable Image Versions
+
+## Summary
+
+When deploying Assisted Installer a ConfigMap is required to do so. In this
+ConfigMap we have the values of various components of Assisted Installer including
+the images for the application. In the code for assisted-service, we define
+several environment variables for these images. We also provide default 
+values for these images which can differ from the desired value at runtime.
+
+For many container based software applications the version of the container ran 
+is determined at runtime. Consider the simple example of `docker run myrepo/myimage` 
+the docker client will resolve this to `hub.docker.com/myrepo/myimage:latest` and 
+then run the image on your system. The expectation is that image ran would be the 
+version you supplied using latest as the default tag.
+
+If we expand this example out to container orchestration technologies, like Openshift,
+the expectation is that the version supplied to Openshift would be the version 
+that is ran for all images and containers. The expectation is that if the version
+you supplied to Openshift doesn't exist then it would fail to pull the image and 
+throw the `ImagePullBackOff` error.
+
+## Motivation
+
+### Goals
+
+The main goal of this proposal is to have the ability to control the images used in 
+an Assisted Installer deployment using standard container deployment methods.
+
+### Non-Goals
+
+There are default versions of Openshift Container Platform and RHCOS images
+supplied to Assisted Installer. These are not the focus of this discussion only
+how we decide which images to deploy and interact with.
+
+## Proposal
+
+First, if we need to populate these environment variables in the code then 
+Assisted Installer should populate them with the environment variables given 
+in the ConfigMap at runtime.
```

Comment:
For the SaaS, or generally non-operator deployments on k8s, I wouldn't worry about explaining any automation that exists. Probably what matters is this, assuming I have the facts right: Environment variables need to be set on the assisted-service Pod, telling it what container images to use, and that is typically accomplished by putting those values in a ConfigMap and referencing the ConfigMap in the Pod definition.

For the operator: it looks like the CSV's Deployment definition has [image references hard-coded as environment variables](https://github.com/openshift/assisted-service/blob/bff1eb2/deploy/olm-catalog/manifests/assisted-service-operator.clusterserviceversion.yaml#L579-L591). Then I assume when the operator deploys the operand, it puts those same values into the same environment variables in a ConfigMap, and references that from the assisted-service Deployment definition? Or does it skip the ConfigMap and hard-code them onto the Deployment definition?

So I think there are two ways to take the above info and make the enhancement more clear:

1. Add the above details to be clear about where the configmap comes into play, and what the current expectation is at deployment time vs. the proposed expectation. There are several references in the proposal to "the ConfigMap", and it would help to clarify exactly what that is.
2. Don't mention configmaps or other mechanisms at all. Instead, just focus on the environment variables as the API. Currently they are optional, and the proposal is for them to be mandatory. Someone or something will have to set them, and it doesn't matter if it's by configmap, in a Pod definition, or any other way.

I'm leaning toward 2. The configmap might just be an implementation detail that isn't central to this proposal. WDYT?

Reply:
To me option 2 makes more sense for this proposal. It wasn't really the intention to try to codify how the different deployment methods would change in this enhancement. Obviously they would need to change but I feel these are implementation details that will need to be ironed out once the environment variables become required.